Correlator PR corresponding to l1t-offline PR 1063

aloeliger opened the pull request:
#### PR description:

This PR now takes the place of #41260 which messed up the git history, and supersedes the original PR #40670

Previous PR description:

> This is an update of the correlator code
> 
> Includes:
> 
> https://github.com/p2l1pfp/cmssw/pull/92 by @thesps and @slaurila
> https://github.com/p2l1pfp/cmssw/pull/94 by @thomreis and @cerminar (only adds a new feature to the emulator, doesn't change any of the existing emulated values)
> https://github.com/p2l1pfp/cmssw/pull/96 by @thesps
> https://github.com/p2l1pfp/cmssw/pull/97 by @drankincms
> https://github.com/p2l1pfp/cmssw/pull/98 by @drankincms (not used by default)
> https://github.com/p2l1pfp/cmssw/pull/100 by @thesps
> https://github.com/p2l1pfp/cmssw/pull/101 by @gpetruc (not used in production)
> https://github.com/p2l1pfp/cmssw/pull/105 by @gpetruc (not used in production)
> https://github.com/p2l1pfp/cmssw/pull/107 by @gpetruc (transparent)
> 
> Local: https://github.com/cms-l1t-offline/cmssw/pull/1063

My apologies to the original authors, but the code has been squashed to remove history. I have added everyone as co-authors to the squashed commit, but it may not be your github email address. 

I will be updating this PR as requested in the original thread and the messed up PR, as soon as I am able.

EmyrClement committed:
Add SC jets from extended tracks to event content.

aloeliger committed:
Correlator PR corresponding to l1t-offline #1063

Co-authored-by: Sioni Summers <sioni.paris.summers@cern.ch>
Co-authored-by: Santeri Laurila <santeri.laurila@cern.ch>
Co-authored-by: Thomas Reis <thomas.reis@cern.ch>
Co-authored-by: Gianluca Cerminara <gianluca.cerminara@cern.ch>
Co-authored-by: Dylan Rankin <dylan.sheldon.rankin@cern.ch>
Co-authored-by: Giovanni Petrucciani <giovanni.petrucciani@cern.ch>
Co-authored-by: Emyr Clement <emyr.john.clement@cern.ch>

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41279/35046

- This PR adds an extra 100KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Configuration/python/L1Trigger_EventContent_cff.py  modified in PR(s): #41258
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #40563

cmsbuild commented:
A new Pull Request was created by  @aloeliger (Andrew Loeliger) for master.

It involves the following packages:

- DataFormats/L1TParticleFlow (**upgrade, l1**)
- L1Trigger/Configuration (**l1**)
- L1Trigger/Phase2L1ParticleFlow (**upgrade, l1**)


@epalencia, @cmsbuild, @AdrianoDee, @srimanob, @aloeliger, @cecilecaillol can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @missirol, @rovere this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

aloeliger committed:
Merge pull request #2 from EmyrClement/bTagNN_addExtSCJets_master

Add SC jets from extended tracks to event content.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41279/35050

- This PR adds an extra 16KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Configuration/python/L1Trigger_EventContent_cff.py  modified in PR(s): #41258
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #40563

cmsbuild commented:
Pull request #41279 was updated. @epalencia, @cmsbuild, @AdrianoDee, @srimanob, @aloeliger, @cecilecaillol can you please check and sign again.

aloeliger committed:
Add tensorflow session to cache, and reference it at ID level

aloeliger commented:
@makortel In reference to: https://github.com/cms-sw/cmssw/pull/40670#discussion_r1120877734, could you look over my shoulder a bit on a22c934 and make sure this solves your comment adequately. This should put the session at the cache level and reference it at the ID level for running of the NN, but I'm tossing around references to pointers, which seems like the sort of thing one should double check.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41279/35052

- This PR adds an extra 104KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Configuration/python/L1Trigger_EventContent_cff.py  modified in PR(s): #41258
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #40563

aloeliger commented:
@srimanob It may be necessary to change some implementation to FileInPath implementation to work with that test? Maybe the authors can comment on that.

makortel commented on L1Trigger/Phase2L1ParticleFlow/src/BJetId.cc:
```diff
@@ -0,0 +1,92 @@
+BJetId::BJetId(const std::string &iInput,
+               const std::string &iOutput,
+               const BJetTFCache *cache,
+               const std::string &iWeightFile,
```
The `iWeightFile` parameter appears to be unused now
```suggestion
```

aloeliger replied:
This has been removed as part of the FileInPath change

makortel commented on L1Trigger/Phase2L1ParticleFlow/plugins/L1BJetPFJetProducer.cc:
I'd suggest to rename the file to `L1BJetProducer.cc` to be consistent with the class name

aloeliger replied:
This has been done

makortel commented on L1Trigger/Phase2L1ParticleFlow/python/L1BJetProducer_cff.py:
```diff
@@ -0,0 +1,23 @@
+
+from L1Trigger.Phase2L1ParticleFlow.l1pfJetMet_cff import L1TPFJetsExtendedTask
+
+l1tBJetProducerPuppi = cms.EDProducer("L1BJetProducer",
```
It would be better to
```suggestion
from L1Trigger.Phase2L1ParticleFlow.L1BJetProducer_cfi import L1BJetProducer as _L1BJetProducer
l1tBJetProducerPuppi = _L1BJetProducer.clone(
    # whatever is changed wrt the default values
)
```
to avoid repeating the defaults, and to make clear what is changed wrt. the defaults.

aloeliger replied:
This has been done

makortel commented on L1Trigger/Phase2L1ParticleFlow/src/CaloClusterer.cc:
```diff
@@ -167,9 +167,15 @@ l1tpf_calo::SingleCaloClusterer::SingleCaloClusterer(const edm::ParameterSet &ps
       precluster_(*grid_),
       clusterIndex_(*grid_),
       cellKey_(*grid_),
+      preciseEtaPhi_(pset.existsAs<bool>("usePreciseEtaPhi") ? pset.getParameter<bool>("usePreciseEtaPhi") : false),
```
Would it be feasible to rely on `fillDescriptions()` instead of calling `existsAs()`?

aloeliger replied:
In theory, yes. But given the complexity of the configuration (L1Trigger/Phase2L1ParticleFlow/python/l1tPFClustersFromCombinedCalo_cfi.py) this is probably best left to the original developers to sort out what an appropriate default is for the individual producers that use this utility, and all the other defaults for the (2?) different producers dependent on it.

gpetruc replied:
it would be possible, but in the near future this class is going do be removed in favour of the new GCT emulator of PF clustering ( https://github.com/cms-l1t-offline/cmssw/pull/1095 ), so it wouldn't be the best investment in terms of effort.

makortel commented on L1Trigger/Phase2L1ParticleFlow/plugins/DeregionizerProducer.cc:
```diff
@@ -170,19 +173,5 @@ void DeregionizerProducer::setRefs_(l1t::PFCandidate &pf, const l1ct::PuppiObjEm
   }
 }
 
-void DeregionizerProducer::fillDescriptions(edm::ConfigurationDescriptions &descriptions) {
```
Why is `fillDescriptions()` removed?

aloeliger replied:
I would hazard a guess that it was removed because of the extremely complicated `linkConfigs` configuration added into the configuration here: https://github.com/cms-sw/cmssw/pull/41279/files#diff-5361519969ac836bcfcd1766c340fc1f82ffdf021ff12943f216334ebc692870R56. I'm willing to bet that there was no simple solution to making a default value for this particular configuration, so the entirety of the description was removed.

makortel replied:
It would be possible to keep the rest and define the `linkConfigs` as "vector of ParameterSets that can contain anything" (combination of https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#ParameterSets_that_Contain_Vecto and https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#Let_the_ParameterSetDescription)

perrotta replied:
@cerminar to speed up the integration, could you please answer why did you remove the fillDeacriptions method here in DeregionizerProducer.cc (and possibly restore it, following perhaps the suggestion of @makortel here above)?

gpetruc replied:
The code change actually from @thesps, but authorship has become tricky after too many rebases in the 6+ months from when it was authored to when it got here. That said, Andrew's guess was most likely right.

If re-adding the method is required for this code to be integrated, I can make a PR to [aloeliger:l1t-1063](https://github.com/aloeliger/cmssw/tree/l1t-1063) to fix it.

aloeliger replied:
@gpetruc Thanks for this. If you would like me to do it, just let me know. Otherwise I will wait for your PR to my branch.

gpetruc replied:
https://github.com/aloeliger/cmssw/pull/3

aloeliger replied:
This should be present now after merging this PR

makortel reviewed (commented):
@aloeliger The TF graph and session looks generally good now. I left some more or less random comments that caught my eye while browsing the code.

smuzaffar commented:
test parameters:
- pull_request = https://github.com/cms-data/L1Trigger-Phase2L1ParticleFlow/pull/5

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/interface/BJetId.h:
```diff
@@ -0,0 +1,46 @@
+  unique_ptr<float[]> fDZ_;
+  unique_ptr<float[]> fDX_;
+  unique_ptr<float[]> fDY_;
+  tensorflow::Session *const &sessionRef_;
```
Well be present in next commit

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/plugins/L1BJetPFJetProducer.cc:
```diff
@@ -0,0 +1,118 @@
+}
+std::unique_ptr<BJetTFCache> L1BJetProducer::initializeGlobalCache(const edm::ParameterSet& cfg) {
+  tensorflow::setLogging("3");
+  std::string lNNFile = cfg.getParameter<std::string>("NNFileName");
+  edm::FileInPath fp(lNNFile);
```
Will be done in next commit

aloeliger committed:
Address (most) comments

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41279/35103

- This PR adds an extra 16KB to repository

- Found files with invalid states:
  - L1Trigger/Phase2L1ParticleFlow/plugins/L1BJetPFJetProducer.cc:
    - Added: 4b4967d4a0bb5b2510055d57843527bbc8b88c51
    - Deleted: 455c7bd54689a7bd7c0fdf151c8ac5d724412fb5

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1TParticleFlow/interface/layer1_emulator.h  modified in PR(s): #41290
  - File  DataFormats/L1TParticleFlow/src/classes_def.xml  modified in PR(s): #41290
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #40563
  - File  L1Trigger/Phase2L1ParticleFlow/interface/conifer.h  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/plugins/L1TCtL2EgProducer.cc  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/plugins/PFClusterProducerFromHGC3DClusters.cc  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer2EG_cff.py  modified in PR(s): #41290

cmsbuild commented:
Pull request #41279 was updated. @epalencia, @cmsbuild, @AdrianoDee, @srimanob, @aloeliger, @cecilecaillol can you please check and sign again.

aloeliger commented:
@thesps @slaurila @thomreis @cerminar @drankincms & @gpetruc you are the authors I could find on this code originally. Before I start any testing here, I'm going to request that you confirm that the code in this branch works under whatever tests you performed for it originally. It has now had to be changed considerably from it's initial form, and changed even more than what got merged into the L1 offline branch before my tenure as software co-coordinator. I would be very skeptical of this software unless it can be tested and demonstrated to work again as the correlator team and original authors intended.

cerminar commented:
@aloeliger thanks for the changes!
I repeat here what I already expressed in another discussion. 

My personal take is that we can not really afford to repeat physics validation during the integration process, If the code review requires changes so big to jeopardize the functionality, there is a problem with the procedure.
Anyhow, the release integration already foresees several tests to catch glitches in the process, among which relval tests. Moreover, we are not discussing a backport to a production release for data-taking. I would simply proceed with the integration, and we will address potential issues in future updates if need be.

aloeliger commented:
please test

srimanob commented:
Hi @aloeliger @cerminar 
Could you please clarify the effect of this PR? Will it be used for MC production for Phase-2 using 13_1?

aloeliger commented:
> Hi @aloeliger @cerminar Could you please clarify the effect of this PR? Will it be used for MC production for Phase-2 using 13_1?

@srimanob Yes, this my understanding. I have been told that several downstream L1T emulators hoping to be used in 13_1 MC production are reliant on the changes here.

srimanob commented:
Following the schedule, we will miss CMSSW_13_1_0_pre3 (2023/04/11), right @perrotta @rappoccio ?

The last open for 13_1 is CMSSW_13_1_0_pre4: 2023/05/02. When it is integrated and we have a release, we should try to produce high stat relvals.

aloeliger commented:
> Following the schedule, we will miss CMSSW_13_1_0_pre3 (2023/04/11), right @perrotta @rappoccio ?
> 
> The last open for 13_1 is CMSSW_13_1_0_pre4: 2023/05/02. When it is integrated and we have a release, we should try to produce high stat relvals.

@srimanob This is my understanding. I am trying to process this through as quickly as possible as there is another Correlator PR reliant upon this one, and then the proposed GT emulator is reliant upon that PR.

perrotta commented:
> Following the schedule, we will miss CMSSW_13_1_0_pre3 (2023/04/11), right @perrotta @rappoccio ?
> 

The build of pre3 was already started earlier this morning

cmsbuild commented:
-1

**Failed Tests**: RelVals RelVals-INPUT 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/summary.html
**COMMIT**: 455c7bd54689a7bd7c0fdf151c8ac5d724412fb5
**CMSSW**: CMSSW_13_1_X_2023-04-11-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/41279/31931/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals

- [23696.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrix-results/23696.0_CloseByPGun_CE_E_Front_120um+2026D95)
```
----- Begin Fatal Exception 12-Apr-2023 12:26:24 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Validating configuration of module: class=L1BJetProducer label='l1tBJetProducerPuppi'
Exception Message:
Illegal parameter found in configuration.  The parameter is named:
 'nParticle'
You could be trying to use a parameter name that is not
allowed for this plugin or it could be misspelled.
----- End Fatal Exception -------------------------------------------------
```
- [23700.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrix-results/23700.0_CloseByPGun_CE_H_Coarse_Scint+2026D95)
```
----- Begin Fatal Exception 12-Apr-2023 12:26:51 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Validating configuration of module: class=L1BJetProducer label='l1tBJetProducerPuppi'
Exception Message:
Illegal parameter found in configuration.  The parameter is named:
 'nParticle'
You could be trying to use a parameter name that is not
allowed for this plugin or it could be misspelled.
----- End Fatal Exception -------------------------------------------------
```
- [23234.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrix-results/23234.0_TTbar_14TeV+2026D94)
```
----- Begin Fatal Exception 12-Apr-2023 12:30:38 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Validating configuration of module: class=L1BJetProducer label='l1tBJetProducerPuppi'
Exception Message:
Illegal parameter found in configuration.  The parameter is named:
 'nParticle'
You could be trying to use a parameter name that is not
allowed for this plugin or it could be misspelled.
----- End Fatal Exception -------------------------------------------------
```
<details>
<summary>Expand to see more relval errors ...</summary>

- [23634.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrix-results/23634.0_TTbar_14TeV+2026D95)
- [23634.911](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrix-results/23634.911_TTbar_14TeV+2026D95_DD4hep)
- [23834.999](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrix-results/23834.999_TTbar_14TeV+2026D95PU_PMXS1S2PR)
</details>



## RelVals-INPUT

- [20834.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrixINPUT-results/20834.0_TTbar_14TeV+2026D88)```20834.0_TTbar_14TeV+2026D88/step2_TTbar_14TeV+2026D88.log```
- [20834.21](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrixINPUT-results/20834.21_TTbar_14TeV+2026D88_ProdLike)```20834.21_TTbar_14TeV+2026D88_ProdLike/step2_TTbar_14TeV+2026D88_ProdLike.log```
- [21034.21](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrixINPUT-results/21034.21_TTbar_14TeV+2026D88PU_ProdLike)```21034.21_TTbar_14TeV+2026D88PU_ProdLike/step2_TTbar_14TeV+2026D88PU_ProdLike.log```
<details>
<summary>Expand to see more relval errors ...</summary>

- [20834.5](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrixINPUT-results/20834.5_TTbar_14TeV+2026D88_pixelTrackingOnly)
- [20834.501](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrixINPUT-results/20834.501_TTbar_14TeV+2026D88_Patatrack_PixelOnlyCPU)
- [20834.9](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrixINPUT-results/20834.9_TTbar_14TeV+2026D88_vectorHits)
- [20834.502](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrixINPUT-results/20834.502_TTbar_14TeV+2026D88_Patatrack_PixelOnlyGPU)
- [20896.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrixINPUT-results/20896.0_CloseByPGun_CE_E_Front_120um+2026D88)
- [20900.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrixINPUT-results/20900.0_CloseByPGun_CE_H_Coarse_Scint+2026D88)
- [21034.114](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrixINPUT-results/21034.114_TTbar_14TeV+2026D88PU_OTInefficiency10PC)
- [23634.21](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrixINPUT-results/23634.21_TTbar_14TeV+2026D95_ProdLike)
- [23634.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrixINPUT-results/23634.0_TTbar_14TeV+2026D95)
- [23634.103](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrixINPUT-results/23634.103_TTbar_14TeV+2026D95Aging3000)
- [23834.21](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrixINPUT-results/23834.21_TTbar_14TeV+2026D95PU_ProdLike)
- [24034.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrixINPUT-results/24034.0_TTbar_14TeV+2026D96)
- [23634.5](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrixINPUT-results/23634.5_TTbar_14TeV+2026D95_pixelTrackingOnly)
- [23696.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrixINPUT-results/23696.0_CloseByPGun_CE_E_Front_120um+2026D95)
- [23634.9](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrixINPUT-results/23634.9_TTbar_14TeV+2026D95_vectorHits)
- [23700.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31931/runTheMatrixINPUT-results/23700.0_CloseByPGun_CE_H_Coarse_Scint+2026D95)
</details>

srimanob commented on L1Trigger/Phase2L1ParticleFlow/python/L1BJetProducer_cfi.py:
```diff
@@ -0,0 +1,14 @@
+                                NNInput = cms.string("input:0"),
+                                NNOutput = cms.string("sequential/dense_2/Sigmoid"),
+                                maxJets = cms.int32(10),
+                                nParticle = cms.int32(10),
```
Should this be `nParticles`?

aloeliger replied:
Yes.

aloeliger committed:
nParticle -> nParticles

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41279/35141

- This PR adds an extra 148KB to repository

- Found files with invalid states:
  - L1Trigger/Phase2L1ParticleFlow/plugins/L1BJetPFJetProducer.cc:
    - Added: 4b4967d4a0bb5b2510055d57843527bbc8b88c51
    - Deleted: 455c7bd54689a7bd7c0fdf151c8ac5d724412fb5

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1TParticleFlow/interface/layer1_emulator.h  modified in PR(s): #41290
  - File  DataFormats/L1TParticleFlow/src/classes_def.xml  modified in PR(s): #41290
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #40563
  - File  L1Trigger/Phase2L1ParticleFlow/interface/conifer.h  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/plugins/L1TCtL2EgProducer.cc  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/plugins/PFClusterProducerFromHGC3DClusters.cc  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer2EG_cff.py  modified in PR(s): #41290

cmsbuild commented:
Pull request #41279 was updated. @epalencia, @cmsbuild, @AdrianoDee, @srimanob, @aloeliger, @cecilecaillol can you please check and sign again.

aloeliger commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/31939/summary.html
**COMMIT**: 6802f37c04cd016355d02f6d1ebf9894a65eca7a
**CMSSW**: CMSSW_13_1_X_2023-04-12-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/41279/31939/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 15 lines to the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 11 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3459609
 - DQMHistoTests: Total failures: 66
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3459521
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-12-1100+add0e4/56496/triggerResults): found differences in 3 / 46 workflows

aloeliger commented:
@cerminar Are these trigger result changes expected?

thesps commented:
@aloeliger I took a look at the differences, since they all concern HT & jet paths. Since we updated the JECs here it makes sense that there could be some differences.

aloeliger commented:
+l1

cerminar commented:
Hello @makortel , @AdrianoDee, @srimanob  nothing has happened to this PR for about 2 weeks. This is blocking other PRs and further development. Can we proceed asap?

cerminar commented:
@perrotta is there anything we can do to speed up this a little?
Thanks!

perrotta commented on L1Trigger/Phase2L1ParticleFlow/python/L1BJetProducer_cff.py:
```diff
@@ -0,0 +1,21 @@
+from L1Trigger.Phase2L1ParticleFlow.L1BJetProducer_cfi import L1BJetProducer as _L1BJetProducer
+
+l1tBJetProducerPuppi = _L1BJetProducer.clone(
+    jets = cms.InputTag("l1tSCPFL1PuppiExtended", ""),
+    maxJets = cms.int32(6),
+    minPt = cms.double(10),
+    vtx = cms.InputTag("l1tVertexFinderEmulator","l1verticesEmulation"),
```
```suggestion
    jets = ("l1tSCPFL1PuppiExtended", ""),
    maxJets = 6,
    minPt = 10,
    vtx = ("l1tVertexFinderEmulator","l1verticesEmulation")
```

aloeliger replied:
Changed in most recent commit

perrotta commented on L1Trigger/Phase2L1ParticleFlow/python/L1BJetProducer_cff.py:
```diff
@@ -0,0 +1,21 @@
+
+
+l1tBJetProducerPuppiCorrectedEmulator = l1tBJetProducerPuppi.clone(
+    jets = cms.InputTag("l1tSCPFL1PuppiExtendedCorrectedEmulator", ""),
```
```suggestion
    jets = ("l1tSCPFL1PuppiExtendedCorrectedEmulator", "")
```

perrotta commented on L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py:
```diff
@@ -469,6 +474,16 @@
 
+
+l1tLayer1Extended = l1tLayer1.clone(
+    pfProducers = cms.VInputTag(
+        cms.InputTag("l1tLayer1BarrelExtended"),
+        cms.InputTag("l1tLayer1HGCalExtended"),
+        cms.InputTag("l1tLayer1HGCalNoTK"),
+        cms.InputTag("l1tLayer1HF")
```
```suggestion
    pfProducers = [ ("l1tLayer1BarrelExtended"), ("l1tLayer1HGCalExtended"), ("l1tLayer1HGCalNoTK"), ("l1tLayer1HF")]
```

aloeliger replied:
Changed in most recent commit

perrotta commented on L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py:
```diff
@@ -286,6 +290,7 @@
     writeRawHgcalCluster = cms.untracked.bool(True)
 )
 
+l1tLayer1HGCalExtended = l1tLayer1HGCal.clone(tracks = cms.InputTag('l1tPFTracksFromL1TracksExtended'))
```
```suggestion
l1tLayer1HGCalExtended = l1tLayer1HGCal.clone(tracks = ('l1tPFTracksFromL1TracksExtended'))
```

aloeliger replied:
Changed in most recent commit

perrotta commented on L1Trigger/Phase2L1ParticleFlow/python/l1pfJetMet_cff.py:
```diff
@@ -1,23 +1,28 @@
 l1tSCPFL1Puppi         = l1tSeedConePFJetProducer.clone()
 l1tSCPFL1PuppiEmulator = l1tSeedConePFJetEmulatorProducer.clone(L1PFObjects = 'l1tLayer2Deregionizer:Puppi')
+l1tSCPFL1PuppiCorrectedEmulator = l1tSeedConePFJetEmulatorProducer.clone(L1PFObjects = 'l1tLayer2Deregionizer:Puppi',
+                                                                     doCorrections = cms.bool(True),
+                                                                     correctorFile = cms.string("L1Trigger/Phase2L1ParticleFlow/data/jecs/jecs_20220308.root"),
+                                                                     correctorDir = cms.string('L1PuppiSC4EmuJets'))
```
```suggestion
                                                                     doCorrections = True,
                                                                     correctorFile = "L1Trigger/Phase2L1ParticleFlow/data/jecs/jecs_20220308.root",
                                                                     correctorDir = "L1PuppiSC4EmuJets")
```

perrotta commented on L1Trigger/Phase2L1ParticleFlow/python/l1pfJetMet_cff.py:
```diff
@@ -26,6 +31,21 @@
+l1tSCPFL1PuppiExtended         = l1tSeedConePFJetProducer.clone(L1PFObjects = 'l1tLayer1Extended:Puppi')
+l1tSCPFL1PuppiExtendedEmulator = l1tSeedConePFJetEmulatorProducer.clone(L1PFObjects = 'l1tLayer2DeregionizerExtended:Puppi')
+l1tSCPFL1PuppiExtendedCorrectedEmulator = l1tSeedConePFJetEmulatorProducer.clone(L1PFObjects = 'l1tLayer2DeregionizerExtended:Puppi',
+                                                                     doCorrections = cms.bool(True),
+                                                                     correctorFile = cms.string("L1Trigger/Phase2L1ParticleFlow/data/jecs/jecs_20220308.root"),
+                                                                     correctorDir = cms.string('L1PuppiSC4EmuJets'))
```
```suggestion
                                                                     doCorrections = True,
                                                                     correctorFile = "L1Trigger/Phase2L1ParticleFlow/data/jecs/jecs_20220308.root",
                                                                     correctorDir = "L1PuppiSC4EmuJets")
```

aloeliger replied:
Changed in most recent commit

perrotta reviewed (commented):
None of the comments below is really urgent. Please apply if some other update is needed, otherwise they could even be deferred to a follow up PR

perrotta commented:
@AdrianoDee, @srimanob please review, or set a time for doing so.

perrotta commented:
There is also an ongoing discussion about fillDescription in https://github.com/cms-sw/cmssw/pull/41279#discussion_r1158665873

I would say that adding back that method is highly preferable than removing it. Then, the less urgent [comments above](https://github.com/cms-sw/cmssw/pull/41279#pullrequestreview-1401619949) could also get implemented

gpetruc committed:
Recover fillDescriptions

AdrianoDee commented:
@perrotta @aloeliger will take care between today and tomorrow.

aloeliger committed:
Merge pull request #3 from gpetruc/l1t-1063

Recover fillDescriptions as requested

aloeliger commented:
> None of the comments below is really urgent. Please apply if some other update is needed, otherwise they could even be deferred to a follow up PR

Since we're waiting and will need to restart the tests anyways, I'll take a look at these ~now.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41279/35277

- This PR adds an extra 16KB to repository

- Found files with invalid states:
  - L1Trigger/Phase2L1ParticleFlow/plugins/L1BJetPFJetProducer.cc:
    - Added: 4b4967d4a0bb5b2510055d57843527bbc8b88c51
    - Deleted: 455c7bd54689a7bd7c0fdf151c8ac5d724412fb5

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1TParticleFlow/interface/layer1_emulator.h  modified in PR(s): #41290
  - File  DataFormats/L1TParticleFlow/src/classes_def.xml  modified in PR(s): #41290
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #41224, #40563
  - File  L1Trigger/Phase2L1ParticleFlow/interface/conifer.h  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/plugins/L1TCtL2EgProducer.cc  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/plugins/PFClusterProducerFromHGC3DClusters.cc  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer2EG_cff.py  modified in PR(s): #41290

cmsbuild commented:
Pull request #41279 was updated. @epalencia, @cmsbuild, @AdrianoDee, @srimanob, @aloeliger, @cecilecaillol can you please check and sign again.

aloeliger commented:
@perrotta Before I forget it, this has required externals in CMSSW data, does that need to be integrated first? Does that have special testing requirements or can I go start it?

aloeliger committed:
Implement non-urgent format recommendations

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/plugins/L1BJetProducer.cc:
```diff
@@ -0,0 +1,116 @@
+class L1BJetProducer : public edm::stream::EDProducer<edm::GlobalCache<BJetTFCache>> {
+public:
+  explicit L1BJetProducer(const edm::ParameterSet&, const BJetTFCache*);
+  ~L1BJetProducer() override;
```
Changed in most recent commit

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/plugins/L1BJetProducer.cc:
```diff
@@ -0,0 +1,116 @@
+  desc.add<edm::InputTag>("vtx", edm::InputTag("L1VertexFinderEmulator", "l1verticesEmulation"));
+  descriptions.add("L1BJetProducer", desc);
+}
+L1BJetProducer::~L1BJetProducer() {}
```
Changed in most recent commit

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/python/l1tDeregionizerProducer_cfi.py:
```diff
@@ -1,10 +1,69 @@
+)
+
+hgcalPosConfig = hgcalConfig_.clone(
+    outputBoard = cms.int32(4)
```
Changed in most recent commit

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/python/l1tPFTracksFromL1Tracks_cfi.py:
```diff
@@ -25,3 +25,16 @@
 )
 
+l1tPFTracksFromL1TracksExtended = l1tPFTracksFromL1Tracks.clone(
+    L1TrackTag = cms.InputTag("l1tTTTracksFromExtendedTrackletEmulation", "Level1TTTracks"),
```
Changed in most recent commit

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/python/l1tPFTracksFromL1Tracks_cfi.py:
```diff
@@ -25,3 +25,16 @@
+    qualityBits = cms.vstring( 
+        "momentum.perp > 2 && getStubRefs.size >= 4 && chi2Red < 15 && POCA.x < 1.0 && POCA.x > -1.0 && POCA.y < 1.0 && POCA.y > -1.0",
+        "momentum.perp > 2 && getStubRefs.size >= 6 && chi2Red < 15 && chi2 < 50 && POCA.x < 1.0 && POCA.x > -1.0 && POCA.y < 1.0 && POCA.y > -1.0", # historical reasons
+        "momentum.perp > 5 && getStubRefs.size >= 4 && POCA.x < 1.0 && POCA.x > -1.0 && POCA.y < 1.0 && POCA.y > -1.0"
```
Changed in most recent commit

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/python/l1tPFTracksFromL1Tracks_cfi.py:
```diff
@@ -25,3 +25,16 @@
+        #using restriction on track x,y for now
+        #long term fix would be to alter track propagation in L1Trigger/Phase2L1ParticleFlow/plugins/PFTrackProducerFromL1Tracks.cc
+    ),
+    redigitizeTrackWord = cms.bool(True),
```
Changed in most recent commit

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/test/make_l1ct_patternFiles_cfg.py:
```diff
@@ -42,8 +42,11 @@
+process.l1tLayer2SeedConeJetsCorrected = l1tSeedConePFJetEmulatorProducer.clone(L1PFObject = cms.InputTag('l1tLayer2Deregionizer', 'Puppi'),
+                                                                                doCorrections = cms.bool(True),
+                                                                                correctorFile = cms.string("L1Trigger/Phase2L1ParticleFlow/data/jecs/jecs_20220308.root"),
+                                                                                correctorDir = cms.string('L1PuppiSC4EmuJets'))
```
Changed in most recent commit

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/test/make_l1ct_patternFiles_fromRAW_cfg.py:
```diff
@@ -41,8 +41,11 @@
+process.l1tLayer2SeedConeJetsCorrected = l1tSeedConePFJetEmulatorProducer.clone(L1PFObject = cms.InputTag('l1tLayer2Deregionizer', 'Puppi'),
+                                                                                doCorrections = cms.bool(True),
+                                                                                correctorFile = cms.string("L1Trigger/Phase2L1ParticleFlow/data/jecs/jecs_20220308.root"),
+                                                                                correctorDir = cms.string('L1PuppiSC4EmuJets'))
```
Changed in most recent commit

aloeliger commented:
please test

cmsbuild commented:
Pull request #41279 was updated. @epalencia, @AdrianoDee, @srimanob, @cecilecaillol, @aloeliger can you please check and sign again.

aloeliger commented:
@AdrianoDee Please also note that this PR goes along with https://github.com/cms-data/L1Trigger-Phase2L1ParticleFlow/pull/5 which I think will need your signature too.

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/32154/summary.html
**COMMIT**: d1aa6837e7facb15f0a293c5a52caa6d4218c2bf
**CMSSW**: CMSSW_13_1_X_2023-04-25-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/41279/32154/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @civanch cms-sw/cmssw#41404
 - @cms-sw cms-sw/cmsdist#8213
 - @cms-sw cms-sw/cmsdist#8469
 - @cms-sw cms-sw/cmsdist#8451

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/32154/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/32154/git-merge-result


## Comparison Summary



**Summary**:
 - You potentially added 33 lines to the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 11 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3460685
 - DQMHistoTests: Total failures: 67
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3460596
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-25-2300+add0e4/56679/triggerResults): found differences in 3 / 46 workflows

aloeliger commented:
+l1

AdrianoDee commented on L1Trigger/Phase2L1ParticleFlow/plugins/L1SeedConePFJetProducer.cc:
```diff
@@ -32,31 +33,37 @@ class L1SeedConePFJetProducer : public edm::global::EDProducer<> {
   unsigned nJets;
   bool HW;
   bool debug;
+  bool doCorrections;
```
This could be `const` (also the others above, technically).

aloeliger replied:
this has been changed

AdrianoDee reviewed (commented):
Should be all.

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/interface/conifer.h:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    std::vector<double> _decision_function_double(std::vector<double> x) const {
+      /* Do the prediction with data in/out as double, cast to T, U before prediction */
```
I can find no other instance of this function in the PR, but conifer is an external BDT firmware system, so I don't really want to mess about in anything that may be external to this PR/the emulation.

gpetruc replied:
@thesps

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/plugins/PFClusterProducerFromL1EGClusters.cc:
```diff
@@ -29,16 +33,34 @@ namespace l1tpf {
 l1tpf::PFClusterProducerFromL1EGClusters::PFClusterProducerFromL1EGClusters(const edm::ParameterSet &iConfig)
```
Seems to largely just be author choice. It doesn't otherwise seem to help or harm organization at all.

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/python/L1BJetProducer_cfi.py:
Just to play devil's advocate, would it not be easier to copy and edit defaults with this (maybe redundant) configuration of the module around?

makortel replied:
> copy and edit defaults

Do you mean copying the `cfi` file around and editing the values in the file, or cloning the module python object and changing some parameters?

aloeliger replied:
I meant cloning the module python object. I understand even in the cloning you can provide parameters that are not explicit because there exists a description default, but from a module comprehensibility standpoint it sort of makes sense to me to at least have one spot in python where all parameters are laid out, even if they are default.

However, I will relay this back to @gpetruc, @thesps and @cerminar for their judgement on whether this is truly necessary.

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/python/L1BJetProducer_cff.py:
```diff
@@ -0,0 +1,21 @@
+
+from L1Trigger.Phase2L1ParticleFlow.l1pfJetMet_cff import L1TPFJetsExtendedTask
+
+from L1Trigger.Phase2L1ParticleFlow.L1BJetProducer_cfi import L1BJetProducer as _L1BJetProducer
```
Same comment as above.

perrotta replied:
L1Trigger/Phase2L1ParticleFlow/python/L1BJetProducer_cfi.py is a file with identical name and content than the one created by the fillDescriptions method, and as such it can only generate confusion, for example when parsing from the release. I agree with @AdrianoDee : please remove this explicit file

perrotta replied:
Here there is nothing to remove: it is a cff which clones from other configurations and it must stay

perrotta replied:
In L1BJetProducer_cff.py there is nothing to remove, instead: it is a cff which clones from other configurations and it must stay

aloeliger replied:
Okay. I can remove it. I'm finalizing the other requests. Should be ready soon~ish.

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/interface/corrector.h:
```diff
@@ -54,12 +58,15 @@ namespace l1tpf {
     std::unique_ptr<TH1> index_;
     std::vector<TGraph *> corrections_;
+    std::vector<TH1 *> correctionsEmulated_;
     bool is2d_;
     unsigned int neta_, nemf_;
     float emfMax_;
+    bool emulate_;
```
They cannot be as there is a `std::swap` defined for these elements as part of the `l1tpf::corrector::operator=` operator

AdrianoDee replied:
Right, sorry for the noise!

AdrianoDee replied:
Right, sorry for the noise.

gpetruc commented on L1Trigger/Phase2L1ParticleFlow/interface/conifer.h:
```diff
@@ -0,0 +1,155 @@
+* Balanced tree reduce implementation.
+* Reduces an array of inputs to a single value using the template binary operator 'Op',
+* for example summing all elements with Op_add, or finding the maximum with Op_max
+* Use only when the input array is fully unrolled. Or, slice out a fully unrolled section
```
i don't agree here
@ @

cerminar replied:
Can we get a certified Language Editor to review the comments?

AdrianoDee replied:
Fair enough.

aloeliger committed:
Add upgrade requests

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/interface/deregionizer/deregionizer_input.h:
```diff
@@ -2,46 +2,67 @@
+    bool debug_ = false;
+    // these are not configurable in current design
+    uint nInputFramesPerBX_ = 9;
+    uint tmuxFactor_ = 6;
```
This has been changed

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/interface/regionizer/multififo_regionizer_ref.h:
```diff
@@ -113,7 +137,10 @@ namespace l1ct {
 
     template <typename T>
-    void fillCaloLinks_(unsigned int iclock, const std::vector<DetectorSector<T>>& in, std::vector<T>& links);
+    void fillCaloLinks_(unsigned int iclock,
```
This has been changed

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/interface/regionizer/buffered_folded_multififo_regionizer_ref.h:
```diff
@@ -0,0 +1,142 @@
+#ifndef buffered_folded_multififo_regionizer_ref_h
```
This has been changed

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/interface/regionizer/folded_multififo_regionizer_ref.h:
```diff
@@ -0,0 +1,89 @@
+#ifndef folded_multififo_regionizer_ref_h
```
this has been changed

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/plugins/PFClusterProducerFromL1EGClusters.cc:
```diff
@@ -18,6 +19,9 @@ namespace l1tpf {
     double etCut_;
+    std::vector<double> etaBounds_;
+    std::vector<double> phiBounds_;
+    std::vector<unsigned int> maxClustersEtaPhi_;
```
this has been changed

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/plugins/L1BJetProducer.cc:
```diff
@@ -0,0 +1,114 @@
+  double fMaxEta_;
+  unsigned int fMaxJets_;
+  int fNParticles_;
+  edm::EDGetTokenT<std::vector<l1t::VertexWord>> fVtxEmu_;
```
this has been changed

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41279/35310

- This PR adds an extra 28KB to repository

- Found files with invalid states:
  - L1Trigger/Phase2L1ParticleFlow/python/L1BJetProducer_cfi.py:
    - Added: 455c7bd54689a7bd7c0fdf151c8ac5d724412fb5
    - Modified: 6802f37c04cd016355d02f6d1ebf9894a65eca7a
    - Deleted: 51266ce34a6e86f28f4806d93a531a7136323739
  - L1Trigger/Phase2L1ParticleFlow/plugins/L1BJetPFJetProducer.cc:
    - Added: 4b4967d4a0bb5b2510055d57843527bbc8b88c51
    - Deleted: 455c7bd54689a7bd7c0fdf151c8ac5d724412fb5

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1TParticleFlow/interface/layer1_emulator.h  modified in PR(s): #41290
  - File  DataFormats/L1TParticleFlow/src/classes_def.xml  modified in PR(s): #41290
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #41224, #40563
  - File  L1Trigger/Phase2L1ParticleFlow/interface/conifer.h  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/plugins/L1TCtL2EgProducer.cc  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/plugins/PFClusterProducerFromHGC3DClusters.cc  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer2EG_cff.py  modified in PR(s): #41290

cmsbuild commented:
Pull request #41279 was updated. @epalencia, @cmsbuild, @AdrianoDee, @srimanob, @aloeliger, @cecilecaillol can you please check and sign again.

aloeliger commented on L1Trigger/Phase2L1ParticleFlow/python/l1tSeedConePFJetProducer_cfi.py:
I have added a description to the module with using as defaults the values in the `cfi`. I have also removed these values from the `cfi`. @cerminar @gpetruc @thesps Would you please confirm you are okay with that change?

gpetruc replied:
I'm not sure this file as is would work, since the python interpreter isn't aware of the fillDescription and so won't find the parameter to replace.

Perhaps the suggestion was to remove the cfi (as scram will generate one from fillDescriptions automatically), and them modify the cff to import the module from the cfi and replace the parameter?

aloeliger replied:
I think it should, although this is incorrect at the moment because it needs to be `HW = cms.bool(True)` not `HW=True` for a parameter that doesn't technically already exist.

makortel replied:
It technically works in this case, but any further clones that might want  to add new parameters need to use explicit types on the right hand side, which then causes any typos to caught later than in python.  Also, defining two modules in one `cfi` file is against our convention. A better way would be to leave the `l1tSeedConePFJetProducer_cfi.py` to be generated by `fillDescriptions()`, and define `l1tSeedConePFJetEmulatorProducer_cfi.py` with the content
```py
from L1Trigger.Phase2L1ParticleFlow.l1tSeedConePFJetProducer_cfi import l1tSeedConePFJetProducer as _l1tSeedConePFJetProducer
l1tSeedConePFJetEmulatorProducer = _l1tSeedConePFJetProducer.clone(HW = True)
```
or make a `cff` file along
```py
from L1Trigger.Phase2L1ParticleFlow.l1tSeedConePFJetProducer_cfi import *
l1tSeedConePFJetEmulatorProducer = l1tSeedConePFJetProducer.clone(HW = True)
```

Although currently the `l1tSeedConePFJetProducer_cfi.py`  file does not get generated because `L1SeedConePFJetProducer::fillDescriptions()` is missing
```cpp
  descriptions.addWithDefaultLabel(desc);
  // or
  descriptions.add("l1tSeedConePFJetProducer", desc);
```

AdrianoDee replied:
> Perhaps the suggestion was to remove the cfi (as scram will generate one from fillDescriptions automatically), and them modify the cff to import the module from the cfi and replace the parameter?

Just to confirm what, now, is clear: yes that's what I meant. Sorry if was not clear enough!

AdrianoDee replied:
> Perhaps the suggestion was to remove the cfi (as scram will generate one from fillDescriptions automatically), and them modify the cff to import the module from the cfi and replace the parameter?

Just to confirm what, now, is clear: yes that's what I meant. Sorry if was not clear enough earlier.

gpetruc commented on L1Trigger/Phase2L1ParticleFlow/interface/conifer.h:
```diff
@@ -1,5 +1,5 @@
-#ifndef CONIFER_CPP_H__
-#define CONIFER_CPP_H__
+#ifndef L1TRIGGER_PHASE2L1PARTICLEFLOW_H
+#define L1TRIGGER_PHASE2L1PARTICLEFLOW_H
```
no wait this change is wrong, as the name should be unique.
It should be something that has both the subsystem and the file, e.g. `L1Trigger_Phase2L1ParticleFlow_Conifer_h`

aloeliger replied:
Ah, this is my naming error, That's a quick fix.

gpetruc reviewed (commented):
Two of the latest changes seem wrong to me.

aloeliger committed:
minor corrections

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41279/35312

- This PR adds an extra 32KB to repository

- Found files with invalid states:
  - L1Trigger/Phase2L1ParticleFlow/python/L1BJetProducer_cfi.py:
    - Added: 455c7bd54689a7bd7c0fdf151c8ac5d724412fb5
    - Modified: 6802f37c04cd016355d02f6d1ebf9894a65eca7a
    - Deleted: 51266ce34a6e86f28f4806d93a531a7136323739
  - L1Trigger/Phase2L1ParticleFlow/plugins/L1BJetPFJetProducer.cc:
    - Added: 4b4967d4a0bb5b2510055d57843527bbc8b88c51
    - Deleted: 455c7bd54689a7bd7c0fdf151c8ac5d724412fb5

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1TParticleFlow/interface/layer1_emulator.h  modified in PR(s): #41290
  - File  DataFormats/L1TParticleFlow/src/classes_def.xml  modified in PR(s): #41290
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #41224, #40563
  - File  L1Trigger/Phase2L1ParticleFlow/interface/conifer.h  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/plugins/L1TCtL2EgProducer.cc  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/plugins/PFClusterProducerFromHGC3DClusters.cc  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer2EG_cff.py  modified in PR(s): #41290

cmsbuild commented:
Pull request #41279 was updated. @epalencia, @cmsbuild, @AdrianoDee, @srimanob, @aloeliger, @cecilecaillol can you please check and sign again.

aloeliger commented:
please test

I am going to start the tests as I think most of the comments are at this point handled.

aloeliger commented:
please abort

perrotta commented on L1Trigger/Phase2L1ParticleFlow/python/L1BJetProducer_cff.py:
```diff
@@ -0,0 +1,20 @@
+
+from L1Trigger.Phase2L1ParticleFlow.l1pfJetMet_cff import L1TPFJetsExtendedTask
+
+l1tBJetProducerPuppi = cms.EDProducer(
+    "L1BJetProducer",
+    jets = ("l1tSCPFL1PuppiExtended", ""),
+    maxJets = 6,
+    minPt = 10,
+    vtx = ("l1tVertexFinderEmulator","l1verticesEmulation")
+)
```
```suggestion
from L1Trigger.Phase2L1ParticleFlow.L1BJetProducer_cfi import L1BJetProducer
l1tBJetProducerPuppi = L1tBJetProducer.clone(
    jets = ("l1tSCPFL1PuppiExtended", ""),
    maxJets = 6,
    minPt = 10,
    vtx = ("l1tVertexFinderEmulator","l1verticesEmulation")
)
```

aloeliger committed:
Fix omitted default descriptions

perrotta commented on L1Trigger/Phase2L1ParticleFlow/python/l1tSeedConePFJetProducer_cfi.py:
```diff
@@ -1,12 +1,6 @@
-                           HW          = cms.bool(False),
-                           debug       = cms.bool(False)
-                         )
+l1tSeedConePFJetProducer = cms.EDProducer("L1SeedConePFJetProducer")
```
```suggestion
from L1Trigger.Phase2L1ParticleFlow.l1tSeedConePFJetProducer_cfi import l1tSeedConePFJetProducer
```
if in https://github.com/cms-sw/cmssw/pull/41279/files#diff-fbe4efdfa3a3a6b083dd6fa21bfe8aa555e61bfa1863c36ccdbb0000f6d4fbd8R218 you have named the cfi created by the `fillDescriptions` method `1tSeedConePFJetProducer`

Otherwise, import from it and clone `1tSeedConePFJetProducer` from that config

perrotta reviewed (commented):
There are quite several places that cannot work now, as it was also pointed out by @gpetruc 
The following ones are a few obvious ones, but please test your code before submitting it once again.

perrotta commented on L1Trigger/Phase2L1ParticleFlow/python/l1tSeedConePFJetProducer_cfi.py:
And of course rename this file as l1tSeedConePFJetProducer_cff.py or somehow else if the fillDescriptions of  "L1SeedConePFJetProducer" already creates one file of that name

aloeliger commented:
Okay. I have added the forgotten description declaration. I have also implemented @perrotta's comments for `L1BjetProducer_cff.py` now importing from the automatically generated cfi. Following @makortel's comments, I have removed `l1tSeedConePFJetProducer_cfi.py` and used the description generated one, now called `l1SeedConePFJetProducer_cfi.py`. I have written an explicit cfi for the emulator version called `l1SeedConePFJetEmulatorProducer_cfi.py` following @makortel's suggestion as well.

I have gone through and changed very instance I could find in this PR where those files and objects were mentioned to reflect their new names and files.

@gpetruc, @cerminar. @thesps I need a way to test your code to make sure it all works still. I've been messing around with this so much that we need to make to make sure that these objects and configurations still function as intended. Is there code or test configurations committed that I can use to quickly make sure this all still functions?

thesps commented:
Thanks @aloeliger , I will look over these recent changes also. 
The test configuration that should probe most of this is [L1Trigger/Phase2L1ParticleFlow/test/make_l1ct_patternFiles_cfg.py](https://github.com/aloeliger/cmssw/blob/l1t-1063/L1Trigger/Phase2L1ParticleFlow/test/make_l1ct_patternFiles_cfg.py)

aloeliger commented:
@thesps please be aware I have not pushed the most recent changes because they need to be validated first. Thank you for the test configuration, I will run it.

aloeliger commented:
@thesps this immediately fails complaining `ModuleNotFoundError: No module named 'Configuration.Geometry.GeometryExtended2026D49Reco_cff'`. This geometry is old and no longer exists in 13_1. Have correlator changes been validated with newer Geometry files?

gpetruc commented:
Hi,

To check things run, you can use the runTheMatrix command from the
description of the PR I had made you

Il Ven 28 Apr 2023, 17:57 Andrew Loeliger ***@***.***> ha
scritto:

> @thesps <https://github.com/thesps> this immediately fails complaining ModuleNotFoundError:
> No module named 'Configuration.Geometry.GeometryExtended2026D49Reco_cff'.
> This geometry is old and no longer exists in 13_1. Have correlator changes
> been validated with newer Geometry files?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/41279#issuecomment-1527769924>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABDNWR2K5R4EKEM4RNYVLA3XDPSIPANCNFSM6AAAAAAWTWHXJA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

aloeliger committed:
Update correlator configuration

aloeliger commented:
Okay. The two changed `cff`'s, have no issues when simply testing loading and configuration via bare python, and `runTheMatrix.py  -l 23234.0` runs all the way through with no complaints, which was suggested by the authors as a testing method.

Just to reiterate:

> ... I have added the forgotten description declaration. I have also implemented @perrotta's comments for L1BjetProducer_cff.py now importing from the automatically generated cfi. Following @makortel's comments, I have removed l1tSeedConePFJetProducer_cfi.py and used the description generated one, now called l1SeedConePFJetProducer_cfi.py. I have written an explicit cfi for the emulator version called l1SeedConePFJetEmulatorProducer_cfi.py following @makortel's suggestion as well.
> 
> I have gone through and changed very instance I could find in this PR where those files and objects were mentioned to reflect their new names and files.

Are there remaining issues to handle before I start any testing?

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41279/35332

- This PR adds an extra 16KB to repository

- Found files with invalid states:
  - L1Trigger/Phase2L1ParticleFlow/python/L1BJetProducer_cfi.py:
    - Added: 455c7bd54689a7bd7c0fdf151c8ac5d724412fb5
    - Modified: 6802f37c04cd016355d02f6d1ebf9894a65eca7a
    - Deleted: 51266ce34a6e86f28f4806d93a531a7136323739
  - L1Trigger/Phase2L1ParticleFlow/plugins/L1BJetPFJetProducer.cc:
    - Added: 4b4967d4a0bb5b2510055d57843527bbc8b88c51
    - Deleted: 455c7bd54689a7bd7c0fdf151c8ac5d724412fb5

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1TParticleFlow/interface/layer1_emulator.h  modified in PR(s): #41290
  - File  DataFormats/L1TParticleFlow/src/classes_def.xml  modified in PR(s): #41290
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #41224, #40563
  - File  L1Trigger/Phase2L1ParticleFlow/interface/conifer.h  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/plugins/L1TCtL2EgProducer.cc  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/plugins/PFClusterProducerFromHGC3DClusters.cc  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py  modified in PR(s): #41290
  - File  L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer2EG_cff.py  modified in PR(s): #41290

cmsbuild commented:
Pull request #41279 was updated. @epalencia, @cmsbuild, @AdrianoDee, @srimanob, @aloeliger, @cecilecaillol can you please check and sign again.

aloeliger commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-add0e4/32275/summary.html
**COMMIT**: 46d37afc85f3603ee59e3f4ccce5ed2f008b7ef5
**CMSSW**: CMSSW_13_1_X_2023-05-01-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/41279/32275/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially added 8 lines to the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 11 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3460877
 - DQMHistoTests: Total failures: 67
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3460788
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-05-01-1100+add0e4/56768/triggerResults): found differences in 3 / 46 workflows

aloeliger commented:
+l1

AdrianoDee commented:
+upgrade
(comments addressed, there's [this](https://github.com/cms-sw/cmssw/pull/41279#discussion_r1179033074) only left but to be honest I can't really say it would be blocking from my side)

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented:
@smuzzafar, do I remember correctly that there was a way to get the output of the static analyzer even if it is not accessible through the usual link in https://github.com/cms-sw/cmssw/pull/41279#issuecomment-1530006085 ?
